### PR TITLE
Add WPS support

### PR DESCRIPTION
**Pull request by stefanos1316**
[Issue #34](https://github.com/eellak/epidose/issues/34)

### CODE REVIEW

File: epidose/device/supervisord.conf [hunk 1]
```diff
@@ -26,10 +26,10 @@ redirect_stdout=true           ; redirect proc stderr to stdout (default false)
 stderr_logfile=/var/log/%(program_name)s ; stderr log path, NONE for none; default AUTO
 stderr_logfile_maxbytes=1MB    ; max # logfile bytes b4 rotation (default 50MB)
 
-[program:watchdog]
-command=/opt/venvs/epidose/bin/python /opt/venvs/epidose/lib/python3.7/site-packages/epidose/device/watchdog_d.py
+[program:update_filter]
```
Comment (dspinellis) [inline epidose/device/supervisord.conf]: Why are you moving programs around?  Please avoid unneeded changes, because they make reviewing move difficult.  If some refactoring is needed, please make a separate PR.

File: epidose/device/wps_scanner_d.sh [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+# Pick up utility functions relative to the script's source code
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL="http://ha-server:5010" . "$UTIL"
+
+# Light up LED lights to indicate network status messages.
+# precondition: WiFi should be turned on.
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Why is the WiFi status a precondition here?

File: epidose/device/wps_scanner_d.sh [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	*) log "The given LED option is not available"return 3;;
+  esac
+  sleep 3
+  run_python device_io --"$1"-off
+}
+
+# Connect to a know network if available,
+# otherwise try to connect to the network with
+# the stongest signal by using WPS.
+# precondition: none
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Doesn't this need WiFi to be on?

File: epidose/device/wps_scanner_d.sh [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+# Returns exit codes based on the return codes
+# of the wps_led_blink function
+wps_connect()
+{
+  network_status=$(ifconfig wlan0 | grep "inet ")
+
+  # Check if device is connected to any network
+  if [ -z "$network_status" ] ; then
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: I suggest to test for the opposite condition and return if the connection is there, to reduce the required nesting.

File: epidose/device/wps_scanner_d.sh [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+  return 0
+}
+
+while : ; do
+  # Wait for the Wifi button to be pressed
+  run_python device_io --wifi-wait
+  wifi_acquire
+  if wps_connect; then
+    validity_time=$(get_filter_validity_age)
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Is this required?  Won't the other daemon handle it?  Can't it create a race condition?  Consider just notifying the other daemon, if needed  e.g. by killing its sleep process, which can be recorded in `/var/run/epidose` with something like `sleep … & ; echo … ; wait …`

File: setup.py [hunk 6]
```diff
@@ -39,10 +39,11 @@
         "deploy": ["make-deb"],
     },
     scripts=[
+        "epidose/device/shutdown_epidose.sh",
```
Comment (dspinellis) [inline setup.py]: Again, please minimize unneeded changes.

**Review by dspinellis — CHANGES_REQUESTED** [review]
Thank you.  Please see the comments.  Some could have been avoided by looking critically at the changes you propose, e.g. at the output of `git diff`.

File: epidose/device/supervisord.conf (hunk 1 above)
Comment (stefanos1316) [reply epidose/device/supervisord.conf]: I thought of sorting them alphabetically, that is why I moved them around.
Ok, I will issue a PR about it later.

File: epidose/device/wps_scanner_d.sh (hunk 2 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Actually, the function wps_connect should have the WiFi turned on, I will change this now.

File: epidose/device/wps_scanner_d.sh (hunk 3 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Changed it.

File: epidose/device/wps_scanner_d.sh (hunk 4 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Done.

File: epidose/device/wps_scanner_d.sh (hunk 5 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: If I have understood correctly, you are suggesting to make the following in the update_filter_d

`sleep "$validity_time" & sleep_pid=$!; echo $sleep_pid > /var/run/epidose_update_filter_sleep ; wait $sleep_pid`

and the wps_scanner_d should kill the sleep_pid after it updates the filter?

`get_new_filter
 kill -9 $(cat /var/run/epidose_update_filter_sleep)
`
Another question is if the wps_scanner should update the cuckoo filter after it expires
or anytime it obtains a network connection?

File: setup.py (hunk 6 above)
Comment (stefanos1316) [reply setup.py]: Will do.

File: epidose/device/wps_scanner_d.sh (hunk 5 above)
Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: 1. Exactly!
2. Good point. The cost of getting it is small, so let's get it while it can.

Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Ok, done. I have submitted a new commit.

File: epidose/device/wps_scanner_d.sh [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+set -e
+
+export APP_NAME=wps_scanner_d
+
+# Location of suspend watchdod led
+SUSPEND_WATCHDOG_LED=/var/lock/epidose/suspend-watchdog-led
+
+# Location of update_filter_d sleep process file
+SLEEP_UPDATE_FILTER_PID=/var/run/epidose_update_filter_sleep
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Define it in` util.sh` and use it on both scripts, right?

File: epidose/device/update_filter_d.sh [hunk 8]
```diff
@@ -44,7 +40,8 @@ wait_till_filter_needed()
   validity_time=$(get_filter_validity_age)
   if [ "$validity_time" -ne 0 ]; then
     log "Sleeping for $validity_time s"
-    sleep "$validity_time"
+    # Sleep until filter becomes invalid and store sleep process id
+    sleep "$validity_time" & sleep_pid=$!; echo $sleep_pid > /var/run/epidose_update_filter_sleep ; wait $sleep_pid
```
Comment (dspinellis) [inline epidose/device/update_filter_d.sh]: Please break this into multiple lines.

File: epidose/device/wps_scanner_d.sh [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+# Light up LED lights to indicate network status messages.
+# precondition: none
+# Returns an exit code of 0 if a network connection is established,
+# while exit code 1 denotes a failed attempt to connect to a WPS network,
+# exit code 2 shows that there are no available WPS networks in the area,
+# and value 3 refers to an invalid LED color option.
+wps_led_blink()
+{
+  run_python device_io --"$1"-on
+  case "$1" in
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh:43]: Very nice!

File: epidose/device/wps_scanner_d.sh [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    return 0
+  fi
+
+  log "Initiating WPS connection"
+ 
+  # Suspend watchdog's LED activities
+  touch "$SUSPEND_WATCHDOG_LED"
+  
+  # Get WPS network with the strongest signal
+  get_network_with_strongest_signal=$(wpa_cli -i wlan0 scan_results | grep WPS | sort -r -k3 | tail -1)
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: Use `network_with_strongest_signal` (without `get_`)

File: epidose/device/wps_scanner_d.sh [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  return "$exit_code"
+}
+
+while : ; do
+  # Wait for the Wifi button to be pressed
+  run_python device_io --wifi-wait
+  wifi_acquire
+  if wps_connect; then
+    log "Updating Cuckoo filter" 
+    if get_new_filter; then
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh]: There's no need (and it is risky due to race conditions) to create a second instance of this function. Simply killing the *sleep* will cause it to be called, no?

**Review by dspinellis — CHANGES_REQUESTED** [review]
This was a pleasure to review!  See a few comments.

File: epidose/device/wps_scanner_d.sh (hunk 7 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Done!

File: epidose/device/update_filter_d.sh (hunk 8 above)
Comment (stefanos1316) [reply epidose/device/update_filter_d.sh]: Ok!

File: epidose/device/wps_scanner_d.sh (hunk 9 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh:43]: Thanks, I am trying to use the DRY method :)

File: epidose/device/wps_scanner_d.sh (hunk 10 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Ok.

File: epidose/device/wps_scanner_d.sh (hunk 11 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: Yes, I have removed the get_new_filter function and now it only kills the sleep process.
However, for some reason the update_filter daemon, after the sleep process that is killed,
it exits without any error message.
Any idea of what the issue might be?

Comment (dspinellis) [reply epidose/device/wps_scanner_d.sh]: To debug this issue I constructed a small test script
```
#!/bin/sh
while : ; do
  sleep 5 &
  pid=$!
  echo $pid
  wait $pid
done
```

This works fine, until you add `set -e` at its beginning.  Checking the `wait` exit code e.g. with `  wait $pid || echo "Killed"` addresses the problem.  You should add an appropriate log statement instead of echo.

Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh]: I had no clue that set -e was causing this!
I have applied all the requested changes and I have submitted them.
Please let me know if more changes are required.

File: epidose/device/update_filter_d.sh [hunk 12]
```diff
@@ -44,7 +40,11 @@ wait_till_filter_needed()
   validity_time=$(get_filter_validity_age)
   if [ "$validity_time" -ne 0 ]; then
     log "Sleeping for $validity_time s"
-    sleep "$validity_time"
+    # Sleep until filter becomes invalid and store sleep process id
+    sleep "$validity_time" &
+    sleep_pid=$!
+    echo "$sleep_pid" > "$SLEEP_UPDATE_FILTER_PID"
+    wait $sleep_pid || log "Received SIGKILL; exiting from sleep state"
```
Comment (dspinellis) [inline epidose/device/update_filter_d.sh]: To make logging better consider
```
if wait ; then
  log "Waking up from sleep; new filter is now required"
else
  log "Killed during sleep; obtaining a new filter"
fi
```

File: epidose/device/wps_scanner_d.sh [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+  touch "$SUSPEND_WATCHDOG_LED"
+  
+  # Get WPS network with the strongest signal
+  network_with_strongest_signal=$(wpa_cli -i wlan0 scan_results | grep WPS | sort -r -k3 | tail -1)
+  if [ -z "$network_with_strongest_signal" ] ; then
+    # If there is no available WPS network in the area, turn on the orange LED light	    
+    wps_led_blink orange
+    exit_code=$?
+
+    # Release watchdog
```
Comment (dspinellis) [inline epidose/device/wps_scanner_d.sh:81]: Release watchdog LED suspension

**Review by dspinellis — CHANGES_REQUESTED** [review]
Two more changes and we're done.

File: epidose/device/wps_scanner_d.sh (hunk 13 above)
Comment (stefanos1316) [reply epidose/device/wps_scanner_d.sh:81]: ok.

File: epidose/device/update_filter_d.sh (hunk 12 above)
Comment (stefanos1316) [reply epidose/device/update_filter_d.sh]: Ok, done!
Please check the new commit.

**Review by dspinellis — APPROVED** [review]
